NExtends/Expressions/ExpressionEqualityComparer.cs

```diff
@@ -0,0 +1,239 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NExtends.Expressions
+{
+	/// <summary>
+	/// source : http://stackoverflow.com/questions/283537/most-efficient-way-to-test-equality-of-lambda-expressions
+	/// </summary>
+	public static class ExpressionEqualityComparer
+	{
+		public static bool Eq(
+			LambdaExpression x,
+			LambdaExpression y)
+		{
+			return ExpressionsEqual(x, y, null, null);
+		}
+
+		public static bool Eq<TSource1, TSource2, TValue>(
+			Expression<Func<TSource1, TSource2, TValue>> x,
+			Expression<Func<TSource1, TSource2, TValue>> y)
+		{
+			return ExpressionsEqual(x, y, null, null);
+		}
+
+		public static Expression<Func<Expression<Func<TSource, TValue>>, bool>> Eq<TSource, TValue>(Expression<Func<TSource, TValue>> y)
+		{
+			return x => ExpressionsEqual(x, y, null, null);
+		}
+
+		private static bool ExpressionsEqual(Expression x, Expression y, LambdaExpression rootX, LambdaExpression rootY)
+		{
+			if (ReferenceEquals(x, y)) return true;
+			if (x == null || y == null) return false;
+
+			var valueX = TryCalculateConstant(x);
+			var valueY = TryCalculateConstant(y);
+
+			if (valueX.IsDefined && valueY.IsDefined)
+				return ValuesEqual(valueX.Value, valueY.Value);
+
+			if (x.NodeType != y.NodeType
+				|| x.Type != y.Type)
+			{
+				if (IsAnonymousType(x.Type) && IsAnonymousType(y.Type))
+					throw new NotImplementedException("Comparison of Anonymous Types is not supported");
+				return false;
+			}
+
+			if (x is LambdaExpression)
+			{
+				var lx = (LambdaExpression)x;
+				var ly = (LambdaExpression)y;
+				var paramsX = lx.Parameters;
+				var paramsY = ly.Parameters;
+				return CollectionsEqual(paramsX, paramsY, lx, ly) && ExpressionsEqual(lx.Body, ly.Body, lx, ly);
+			}
+			if (x is MemberExpression)
+			{
+				var mex = (MemberExpression)x;
+				var mey = (MemberExpression)y;
+				return Equals(mex.Member, mey.Member) && ExpressionsEqual(mex.Expression, mey.Expression, rootX, rootY);
+			}
+			if (x is BinaryExpression)
+			{
+				var bx = (BinaryExpression)x;
+				var by = (BinaryExpression)y;
+				return bx.Method == @by.Method && ExpressionsEqual(bx.Left, @by.Left, rootX, rootY) &&
+					   ExpressionsEqual(bx.Right, @by.Right, rootX, rootY);
+			}
+			if (x is UnaryExpression)
+			{
+				var ux = (UnaryExpression)x;
+				var uy = (UnaryExpression)y;
+				return ux.Method == uy.Method && ExpressionsEqual(ux.Operand, uy.Operand, rootX, rootY);
+			}
+			if (x is ParameterExpression)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this type-check-and-cast sequence with an 'as' and a null check. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3247)